lib/internal/modules/esm/hooks.js

```diff
@@ -110,37 +110,27 @@ class Hooks {
   // Cache URLs we've already validated to avoid repeated validation
   #validatedUrls = new SafeSet();
 
-  constructor(userLoaders) {
-    this.addCustomLoaders(userLoaders);
-  }
-
   /**
    * Collect custom/user-defined module loader hook(s).
    * After all hooks have been collected, the global preload hook(s) must be initialized.
-   * @param {import('./loader.js).KeyedExports} customLoaders Exports from user-defined loaders
-   *  (as returned by `ModuleLoader.import()`).
+   * @param {string} url Custom loader specifier
+   * @param {Record<string, unknown>} exports
    */
-  addCustomLoaders(customLoaders = []) {
-    for (let i = 0; i < customLoaders.length; i++) {
-      const {
-        exports,
-        url,
-      } = customLoaders[i];
-      const {
-        globalPreload,
-        resolve,
-        load,
-      } = pluckHooks(exports);
+  addCustomLoader(url, exports) {
+    const {
+      globalPreload,
+      resolve,
+      load,
+    } = pluckHooks(exports);
 
-      if (globalPreload) {
-        ArrayPrototypePush(this.#chains.globalPreload, { fn: globalPreload, url });
-      }
-      if (resolve) {
-        ArrayPrototypePush(this.#chains.resolve, { fn: resolve, url });
-      }
-      if (load) {
-        ArrayPrototypePush(this.#chains.load, { fn: load, url });
-      }
+    if (globalPreload) {
+      ArrayPrototypePush(this.#chains.globalPreload, { fn: globalPreload, url });
+    }
+    if (resolve) {
+      ArrayPrototypePush(this.#chains.resolve, { fn: resolve, url });
+    }
+    if (load) {
+      ArrayPrototypePush(this.#chains.load, { fn: load, url });
     }
```

Comment:
```suggestion
      ArrayPrototypePush(this.#chains.globalPreload, { __proto__: null, fn: globalPreload, url });
    }
    if (resolve) {
      ArrayPrototypePush(this.#chains.resolve, { __proto__: null, fn: resolve, url });
    }
    if (load) {
      ArrayPrototypePush(this.#chains.load, { __proto__: null, fn: load, url });
    }
```

Reply:
I think here this doesn't matter because we only pluck these keys by name.

Reply:
Which means a missing key would fall back to Object.prototype. It only doesn’t matter if you check for own-ness before plucking the key.

Reply:
How would a key be missing?

Reply:
ah, i suppose if every object has `fn` and `url` every time then it's fine, but why leave a time bomb for a future developer?

Reply:
🤷‍♂️ couldn't hurt 👍

And might be easier to just always do this—we happen to know that this isn't necessarily here, but knowing when it is necessary is not common knowledge, and maybe isn't even particularly important knowledge (_someone_ needs to know, but not most contributors). Larger discussion though 😅

Reply:
It could hurt perf, using null prototype changes the optimization path on V8 side – or so I've heard, I don't know if it's still true.
But in this case, I don't see how it can be a "time bomb": the objects are created all at the same place, and having undefined values would crash the process anyway. Anyway, it's off-topic for this PR, I don't think it's necessary and feel confident that it's not going to cause issues.
(`ArrayPrototypePush` also falls back to `%Object.prototype%` when setting the new value(s), and everyone is fine with it 🤷‍♂️)

Reply:
Both fair points. Feel free to resolve this.